src/phoenix/experimental/evals/long_context_evaluators.py

```diff
@@ -0,0 +1,58 @@
+from typing import List, Optional
+
+from phoenix.experimental.evals import PromptTemplate
+from phoenix.experimental.evals.models import BaseEvalModel
+
+
+class Refiner:
+    def __init__(
+        self,
+        model: BaseEvalModel,
+        initial_prompt_template: PromptTemplate,
+        refine_prompt_template: PromptTemplate,
+        reduce_prompt_template: Optional[PromptTemplate] = None,
+    ) -> None:
+        self._model = model
+        self._initial_prompt_template = initial_prompt_template
+        self._refine_prompt_template = refine_prompt_template
+        self._reduce_prompt_template = reduce_prompt_template
+
+    def evaluate(self, chunks: List[str]) -> str:
+        if len(chunks) < 2:
+            raise ValueError("Must provide two or more documents")
+        model = self._model
+        initial_prompt = self._initial_prompt_template.format({"chunk": chunks[0]})
+        accumulator = model(initial_prompt)
+        for chunk in chunks[1:]:
+            refine_prompt = self._refine_prompt_template.format(
+                {"accumulator": accumulator, "chunk": chunk}
+            )
+            accumulator = model(refine_prompt)
+        if not self._reduce_prompt_template:
+            return accumulator
```

Comment:
were your thoughts that "railing" code would live outside of this code? totally makes sense, just trying to understand the scope of responsibility for this code.

Reply:
That's a good question. I think railing probably lives outside of both of these class eventually. If someone wants to use the output of `Refiner` for the purpose of classification, it would make sense to me to chain together a `Refiner` with some kind of classifier class that knows how to rail output. `MapReducer` is another matter, since the "reduce" step would ideally be the classification step. Maybe there is a need for a `MapReduceClassifier`.